pilot/pkg/proxy/envoy/gateway.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2017 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package envoy
+
+import (
+	"strings"
+
+	"reflect"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	routing "istio.io/api/routing/v1alpha2"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+)
+
+func buildGatewayListeners(mesh *meshconfig.MeshConfig, instances []*model.ServiceInstance, config model.IstioConfigStore, ingress model.Node) Listeners {
+	gateways, err := config.List(model.Gateway.Type, model.NamespaceAll)
```

Comment:
nit: IIRC, none of the code in proxy/envoy/ has the concept of namespaces except for ingress.go which needs cleanup. Essentially, the consumer supplies list of route rules to these functions.

Reply:
Secondly, PTAL at the last PR to gateway spec. I added the notion of workload selector in the gateway. So, you can filter the gateways that apply to this workload by instance labels. This won't matter in CF, but will matter in k8s.